File: src/ewoc_cmdln.cc
/**
 * @file    ewoc_cmdl.cc
 *
 * @brief   A command line utility file for subjet energy-weighted observable correlators.
 *          Version 0.2
 */

#include <string>
#include <iostream>

#include "../include/ewoc_cmdln.h"

// =====================================
// Command Line Basics
// =====================================


// ---------------------------------
// Large text
// ---------------------------------
std::string flag_text = R"(
     ______           _____                         _____                _____
 ___|\     \         |\    \   _____           ____|\    \           ___|\    \
|     \     \        | |    | /    /|         /     /\    \         /    /\    \
|     ,_____/|       \/     / |    ||        /     /  \    \       |    |  |    |
|     \--'\_|/       /     /_  \   \/       |     |    |    |      |    |  |____|
|     /___/|        |     // \  \   \       |     |    |    |      |    |   ____
|     \____|\       |    |/   \ |    |      |\     \  /    /|      |    |  |    |
|____ '     /|      |\ ___/\   \|   /|      | \_____\/____/ |      |\ ___\/    /|
|    /_____/ |      | |   | \______/ |       \ |    ||    | /      | |   /____/ |
|____|     | /       \|___|/\ |    | |        \|____||____|/        \|___|    | /
  \( |_____|/           \(   \|____|/            \(    )/             \( |____|/
   '    )/               '      )/                '    '               '   )/
        '                       '                                          '
)";


std::string help_text = R"(
###################################
# EWOC Parameters:
###################################

  # ================================================
  # Basic Options (Required):
  # ================================================
  # Event Generation
  [<--n_events|-n> <int>] :                     Number of generated events;
  [<--level|-l> <str: parton|hadron>] :         QCD level of generated events (parton or hadron);
  [<--process|-p> <str>] :                      Process for generated events (quark, gluon, qcd, w);

  # Jet and Subjet information
  [<--jet_alg|-j> <int>] :                      Jet algorithm (0 [kt], 1 [ca], 2 [antikt]);
  [<--sub_alg|-s> <int>] :                      Subjet algorithm (0 [kt], 1 [ca], 2 [antikt]);
  [<--jet_rad|-R> <double>] :                   Jet radius (can be 'inf' or 'infty');
  [<--sub_rad|-r> <double>] :                   Subjet radius;

  # ================================================
  # Basic Options (Optional):
  # ================================================
  # Phase Space constraints
  [<--pt_min> <double>] :                       Minimum value of jet p_T;
  [<--pt_max> <double>] :                       Maximum value of jet p_T;

  # Misc. Options
  [<--verbose|-v> <int>] :                      Verbosity;


  # ================================================
  # Advanced Options (Optional):
  # ================================================
  # Fragmentation Temperature
  [<--frag_temp|-T> <double>] :                 Temperature associated with string
                                                fragmentation;

  # Coming soon!
  # Rope related parameters


  # ================================================
  # Help:
  # ================================================
  [<--help|-h> ] :                                      Get this help message!


  # ================================================
  # Example commands:
  # ================================================

  # ---------------------------------
  # Single Subjet Radius
  # ---------------------------------
  # Generate EWOC data for 5000 parton-level qcd events
  #R=1 anti-kt jets, r=.1 kt subjets
  ```
    ./pythia_ewocs -n 5000 -l parton -p qcd -j 2 -s 0 -R 1.0 -r 0.1 --pt_min 50 --pt_max 3000

    cd python
    python3 plot_pythia_ewocs.py -n 5000 -l parton -p qcd -j 2 -s 1 -R 1.0 -r 0.1 --pt_min 50 --pt_max 300
    cd ..
  ```

  # ---------------------------------
  # Multiple Subjet Radii
  # ---------------------------------
  # Generate EWOC data for 50000 parton-level qcd events
  # R=1 anti-kt jets, r in [.0,.1,.2,.3,.4,.5] kt subjets
  ```
    for rsub in 0.0 0.1 0.2 0.3 0.4 0.5
      ./pythia_ewocs -n 50000 -l parton -p qcd -j 2 -s 1 -R 1.0 -r $rsub --pt_min 50 --pt_max 3000
    end

    cd python
    python3 plot_pythia_ewocs.py -n 50000 -l parton -p qcd -j 2 -s 1 -R 1.0 -r [0.0,0.1,0.2,0.3,0.4,0.5] --pt_min 50 --pt_max 3000
    cd ..
  ```

  # See the scripts folder and the python folder for event generation and plotting with more options
  ```
  ls scripts/*
  ls python/*.sh
  ```
)";


// ---------------------------------
// Command Line Parameter Verification
// ---------------------------------

/**
* @brief: checks to ensure that valid command line input was given.
*
* @param: argc  Number of command line inputs;
* @param: argv  Vector of command line inputs.
*
* @return: int  1 if failed, 0 if successful
*/
int checkCmdInputs(int argc, char* argv[]) {
    std::cout << flag_text << "\n\n";

    // ---------------------------------
    // Getting command line variables
    // ---------------------------------
    // Basic Settings
    int         n_events      = nevents_cmdln(argc, argv);
    std::string qcd_level     = level_cmdln(argc, argv);
    std::string process_str   = process_cmdln(argc, argv);
    int         jet_alg_int   = jetalg_cmdln(argc, argv);
    double      jet_rad       = jetrad_cmdln(argc, argv);
    int         sub_alg_int   = subalg_cmdln(argc, argv);
    double      sub_rad       = subrad_cmdln(argc, argv);

    // Optional Settings
    double      pt_min        = ptmin_cmdln(argc, argv);
    double      pt_max        = ptmax_cmdln(argc, argv);

    double      E_cm          = Ecm_cmdln(argc, argv);
    std::string s_channel     = schannel_cmdln(argc, argv);

    int         verbose       = verbose_cmdln(argc, argv);

    // Advanced Settings
    double      frag_temp     = fragtemp_cmdln(argc, argv);

    // ---------------------------------
    // Help
    // ---------------------------------
    for(int iarg=0;iarg<argc;iarg++) {
        if(strcmp(argv[iarg], "-h") == 0 or
          strcmp(argv[iarg], "--help") == 0) {
            std::cout << help_text;
            return 1;
        }
    }

    // ---------------------------------
    // Checks on input parameters
    // ---------------------------------
    // If we don't have enough arguments
    if (argc < 19) {
        std::cout << "Missing required command line arguments "
        << "(received " << argc-1 << ").\n"
        << "Use ```--help``` for more detail.\n";
        return 1;
    }

    // Event generator checks
    if (n_events <= 0){
        std::cout << "Number of events must be a positive integer.\n";
        return 1;
    }

    if (qcd_level != "parton" and qcd_level != "hadron") {
        std::cout << "Invalid qcd level " << qcd_level << ", must be 'parton' or 'hadron'.\n";
        return 1;
    }

    if (process_str != "quark" and process_str != "gluon"
      and process_str != "qcd" and process_str != "w") {
        std::cout << "Invalid process " << process_str << ", must be one of (quark, gluon, qcd, w).\n";
        return 1;
    }

    if (jet_alg_int != 0 and jet_alg_int != 1 and jet_alg_int != 2){
        std::cout << "Jet algorithm must be 0 (kt), 1 (ca), or 2 (antikt).\n";
        return 1;
    }

    if (jet_rad <= 0){
        std::cout << "Jet radius must be positive.\n";
        return 1;
    }

    if (sub_alg_int != 0 and sub_alg_int != 1 and sub_alg_int != 2){
        std::cout << "Subjet algorithm must be 0 (ca), 1 (kt), or 2 (antikt).\n";
        return 1;
    }

    if (sub_rad < 0){
        std::cout << "Subjet radius must be positive or zero.\n";
        return 1;
    }

    if (pt_min <= 0 and pt_min != -1){
        std::cout << "pt_min must be positive.\n";
        return 1;
    }

    if (pt_max <= pt_min){
        std::cout << "pt_max must be larger than pt_min.\n";
        return 1;
    }

    if (E_cm != _Ecm_default and E_cm <= 0){
        std::cout << "Center of mass energy must be positive.\n";
        return 1;
    }

    if (s_channel != _schannel_default) {
        if (process_str != "qcd") {
            std::cout << "No s channel production for desired process ("
               << process_str << ").";
            return 1;
        }
        else if (s_channel != "gm" and s_channel != "Z") {
            std::cout << "Invalid s channel option: " << s_channel;
            return 1;
        }
    }

    // Verbosity
    if (verbose >= 1) {
        std::cout << "Function call:\n    ";
        while( argc-- ) std::cout << *(argv++) << " ";
        std::cout << "\n\n";
    }

    // Advanced Options
    if (frag_temp != _frag_temp_default and qcd_level != "hadron") {
        std::cout << "Thermal string fragmentation "
                  << "(given as " << std::to_string(frag_temp)
                  << ") requires hadron level event generation.\n";
        return 1;
    }

    // Return 0 if all checks passed
    return 0;
}


/**
* @brief: Set of functions that return arguments from command
*         line input.
*
* @param: argc/argv         Command line input.
*
* @return: variable type    The desired argument.
*/
// =====================================
// Command Line Reading Utilities
// =====================================

// ---------------------------------
// Required Basic Options
// ---------------------------------

// Number of events
int nevents_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        if(strcmp(argv[iarg], "-n") == 0 or
          strcmp(argv[iarg], "--n_events") == 0) {
            return atoi(argv[iarg+1]);
        }
    }
    return -1;
}

// QCD Level
std::string level_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        // Event Generator Parameters
        if(strcmp(argv[iarg], "-l") == 0 or
          strcmp(argv[iarg], "--level") == 0) {
            return argv[iarg+1];
        }
    }
    return "No level";
}

// Process
std::string process_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        if(strcmp(argv[iarg], "-p") == 0 or
          strcmp(argv[iarg], "--process") == 0) {
            return argv[iarg+1];
        }
    }
    return "No process";
}


// Jet and subjet algorithm
int jetalg_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        if(strcmp(argv[iarg], "-j") == 0 or
          strcmp(argv[iarg], "--jet_alg") == 0) {
            return atoi(argv[iarg+1]);
        }
    }
    return -1;
}

int subalg_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        if(strcmp(argv[iarg], "-s") == 0 or
          strcmp(argv[iarg], "--sub_alg") == 0) {
            return atoi(argv[iarg+1]);
        }
    }
    return -1;
}


// Jet and subjet radii
double jetrad_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        if(strcmp(argv[iarg], "-R") == 0 or
          strcmp(argv[iarg], "--jet_rad") == 0) {
            // Allowing infinite jet radius (whole event)
            if(strcmp(argv[iarg+1], "inf") == 0 or
             strcmp(argv[iarg+1], "infty") == 0 )
                return 1000;
            // Otherwise, we assume it is finite
            return atof(argv[iarg+1]);
        }
    }
    return -1.0;
}

double subrad_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        if(strcmp(argv[iarg], "-r") == 0 or
          strcmp(argv[iarg], "--sub_rad") == 0) {
            return atof(argv[iarg+1]);
        }
    }
    return -1;
}

// ---------------------------------
// Optional Basic Options
// ---------------------------------
// Phase space options
double ptmin_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        // Event Generator Parameters
        if(strcmp(argv[iarg], "--pt_min") == 0)
            return atof(argv[iarg+1]);
    }
    return -1;
}

double ptmax_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        // Event Generator Parameters
        if(strcmp(argv[iarg], "--pt_max") == 0)
            return atof(argv[iarg+1]);
    }
    return std::numeric_limits<double>::max();
}


double _Ecm_default = -1;
double Ecm_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        // Event Generator Parameters
        if(strcmp(argv[iarg], "--E_cm") == 0
         or strcmp(argv[iarg], "-E") == 0)
            return atof(argv[iarg+1]);
    }
    return _Ecm_default;
}

std::string _schannel_default = "gmZ";
std::string schannel_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        // Event Generator Parameters
        if(strcmp(argv[iarg], "--s_channel") == 0)
            return argv[iarg+1];
    }
    return _schannel_default;
}

// Misc. Options
int verbose_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        // Event Generator Parameters
        if(strcmp(argv[iarg], "--verbose") == 0
         or strcmp(argv[iarg], "-v") == 0)
            return atoi(argv[iarg+1]);
    }
    return 0;
}


// ---------------------------------
// Advanced Options
// ---------------------------------
// Thermodynamical Fragmentation
double _frag_temp_default = -10;
double fragtemp_cmdln(int argc, char* argv[]) {
    for(int iarg=0;iarg<argc;iarg++) {
        // Event Generator Parameters
        if(strcmp(argv[iarg], "-T") == 0
         or strcmp(argv[iarg], "--frag_temp") == 0)
            return atof(argv[iarg+1]);
    }
    return _frag_temp_default;
}

// Rope Hadronization
// Coming soon!
double _ropeshove_amp_default = -10;

// Others
// Coming soon!

File: plot_tools/utils/ewoc_cmdln.py
import argparse

from matplotlib.pyplot import legend, show

from sys import exit

from utils.plot_utils import aestheticfig, text_to_hist

from utils.qcd_utils import alg_to_string


# =====================================
# EWOC filenames 
# =====================================

def process_folder(qcd_level, process_str, **kwargs):
    """
    Returns a folder used to store process information given
    a set of command line inputs

    Parameters
    ----------
        Command line arguments; see ewoc_cmdln.py

    Returns
    -------
        string : The relevant folder.

    """
    proc_folder = "output/" + process_str + "_" + qcd_level
    if kwargs['s_channel'] != default_args['s_channel']:
        proc_folder += "_schan_" + kwargs['s_channel']
    proc_folder += "/"
    return proc_folder


def ewoc_folder(n_events, qcd_level, process_str,
        jet_alg_int, jet_rad, *args, **kwargs):
    """
    Returns a folder used to store subjet pair information given
    a set of command line inputs

    Parameters
    ----------
        Command line arguments; see ewoc_cmdln.py

    Returns
    -------
        string : The relevant folder.

    """
    return process_folder(qcd_level, process_str, **kwargs)\
        + f"jetR{float(jet_rad):.1f}/".replace(".", "-")\
        + alg_to_string(jet_alg_int, latex=False) + "jet/"


def ewoc_file_label(n_events, qcd_level, process_str,
        jet_alg_int, jet_rad, sub_alg_int, sub_rad,
        **kwargs):
    """
    Returns a file used to store subjet pair information given
    a set of command line inputs

    Parameters
    ----------
        Command line arguments; see ewoc_cmdln.py

    Returns
    -------
        string : The relevant file.
    """
    folder = ewoc_folder(n_events, qcd_level, process_str,
                jet_alg_int, jet_rad, sub_alg_int, sub_rad,
                **kwargs)
    filename = f"subR{float(sub_rad):.2f}".replace(".", "-") + "_"\
        + alg_to_string(sub_alg_int, latex=False) + "sub"\
        + "_" + str(n_events) + "evts"

    for key in arg_to_str_dict.keys():
        if key in kwargs.keys():
            if kwargs[key] != default_args[key] and kwargs[key] is not None: 
                filename += '_' + arg_to_str_dict[key]\
                         + f"{float(kwargs[key]):.1f}".replace(".", "-")

    return folder + filename


# =====================================
# Command Line Parsing 
# =====================================

# Setup
ewoc_parser=argparse.ArgumentParser('Plot EWOC data from Pythia and FastJet')

# ---------------------------------
# Additional Actions
# ---------------------------------

# Action class for reading jet algorithm information
class parse_jet_alg(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        if values in ['akt', 'antikt', '2', 2]:
            alg_int = 2
        elif values in ['ca', '1', 1]:
            alg_int = 1
        elif values in ['kt', '0', 0]:
            alg_int = 0
        else:
            raise AttributeError("Invalid jet algorithm value: {values}")
        setattr(namespace, self.dest, alg_int)


# Plot pT associated with a particular PID
class plot_pid_pT(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        pt_filename = process_folder(**vars(namespace))\
            + str(values) + "-pt-spectrum.txt" 
        text_to_hist(pt_filename,
                     xlabel=fr"$p^{{\rm PID = {values}}}_T$",
                     nbins=2500, binspace='lin')
                     #nbins=25, binspace='log')
        if vars(namespace)['show']:
            show()
        exit()


# Plot fraction of energy given to softer subjet
class plot_soft_energy(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        kwargs = vars(namespace) 

        for rad_type in ['sub_rad', 'jet_rad']:
            if not isinstance(kwargs[rad_type], list):
                kwargs[rad_type] = [kwargs[rad_type]]

        fig, ax = aestheticfig(ratio_plot=False, showdate=False,
                               xlabel=r"$E_{\rm soft\ subjet}$")

        def soft_energy(z1, z2, Etot):
            z1, z2, Etot = float(z1), float(z2), float(Etot)
            return Etot * min(z1, z2)

        def valid_row(row_info):
            return row_info[0] == 'SP'

        for jet_rad in kwargs['jet_rad']:
            for sub_rad in kwargs['sub_rad']:
                filename = ewoc_file_label(**dict(kwargs,
                                          jet_rad=jet_rad,
                                          sub_rad=sub_rad))
                filename += '.txt'
                text_to_hist(filename, use_cols=[1, 2, 3],
                             col_func=soft_energy,
                             use_rows=valid_row,
                             fig=fig, ax=ax)
        if vars(namespace)['show']:
            show()
        exit()


# Plot pT associated with jets 
class plot_jet_pT(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        fig, ax = aestheticfig(ratio_plot=False, showdate=False,
                               xlabel="Jet Property (GeV)")

        # Finding pT spectrum directly
        pt_filename = ewoc_folder(**vars(namespace))\
            + "jet-pt-spectrum"\
            + "_min"+str(round(vars(namespace)['pt_min']))\
            + "_max"+str(round(vars(namespace)['pt_max']))\
            + ".txt" 
        text_to_hist(pt_filename,
                     xlabel=r"$p^{\rm jet}_T$",
                     nbins=25, binspace='lin',
                     upper_x=2200,
                     label=r"$p_T$ Spectrum",
                     color="steelblue",
                     fig=fig, ax=ax)

        # As a common sense check, finding it independently
        # jet info in the relevant EWOC file
        def valid_row(row_info):
            return row_info[0] == 'J'

        filename = ewoc_file_label(**vars(namespace))
        filename += '.txt'

        text_to_hist(filename, use_cols=[1],
                     use_rows=valid_row,
                     nbins=25, binspace='lin',
                     upper_x=2200,
                     label=r"Energy Spectrum",
                     color="lightcoral",
                     fig=fig, ax=ax)
        legend()

        if vars(namespace)['show']:
            show()
        exit()



# ---------------------------------
# Basic Options (Required)
# ---------------------------------
# Event generation
ewoc_parser.add_argument("-n", "--n_events",
            dest="n_events", type=int,
            help="Number of generated events;",
            required=True)
ewoc_parser.add_argument("-l", "--level", 
            dest="qcd_level", type=str,
            help="QCD level of generated events (parton or hadron)",    
            required=True)
ewoc_parser.add_argument("-p", "--process_str", 
            dest="process_str", type=str,
            help="Process for generated events (quark, gluon, qcd, w);",     
            required=True)

# Jet and subjet information
ewoc_parser.add_argument("-j", "--jet_alg", 
            action=parse_jet_alg, dest="jet_alg_int",
            default=2,
            help="Jet algorithm: 0 [kt], 1 [ca], or "\
                 "2 [akt, antikt] (default);")
ewoc_parser.add_argument("-s", "--sub_alg", 
            action=parse_jet_alg, dest="sub_alg_int",
            default=1,
            help="Subjet algorithm (0 [kt], 1 [ca] (default), "\
                 +"or 2 [akt, antikt];")

ewoc_parser.add_argument("-R", "--jet_rad", 
            nargs='+', type=float,
            dest="jet_rad", 
            help="Jet radius;", default=1.0)
ewoc_parser.add_argument("-r", "--sub_rad", 
            nargs='+', type=float,
            dest="sub_rad",
            help="Subjet radius (or radii);", default=0.1)

# ---------------------------------
# Info for optional variables
# ---------------------------------

default_args    = {'pt_min':    0,
                   'pt_max':    1e5,
                   'E_cm':      4000,
                   'temp':      None,
                   's_channel': 'gmZ'}

arg_to_str_dict = {'pt_min': 'ptmin',
                   'pt_max': 'ptmax',
                   'E_cm':   'Ecm',
                   'temp':   'temp',}

# ---------------------------------
# Basic Options (Optional)
# ---------------------------------
# Phase Space constraints
ewoc_parser.add_argument("--pt_min", 
            dest="pt_min", type=float,
            help="Minimum value of jet p_T;",
            default=default_args['pt_min'],
            required=False)
ewoc_parser.add_argument("--pt_max", 
            dest="pt_max", type=float,
            help="Maximum value of jet p_T;",
            default=default_args['pt_max'],
            required=False)

ewoc_parser.add_argument("--E_cm", "-E", 
            dest="E_cm", type=float,
            help="Center of mass energy;",
            default=default_args['E_cm'],
            required=False)

ewoc_parser.add_argument("--s_channel", 
            dest="s_channel", type=str,
            help="Allowed s-channel mediators;",
            default=default_args['s_channel'],
            required=False)


# ---------------------------------
# Advanced Options (Optional)
# ---------------------------------
ewoc_parser.add_argument("-T", "--frag_temp", 
            dest="temp", type=float, nargs='+',
            help="Temperature of string fragmentation, off by default; see:"\
             +"\n    * https://pythia.org/latest-manual/Fragmentation.html#anchor25;"\
             +"\n    * https://arxiv.org/pdf/1610.09818.pdf;",
            default=default_args['temp'],
            required=False)


# ---------------------------------
# Easy Plotting Options (Optional)
# ---------------------------------

ewoc_parser.add_argument("--plot_type",
            dest="plot_type", type=str,
            help="Argument which sets the type of plot to create."\
                 +"Can be:\n*'sub_rad';\n'pvh';\n*'thermal';",
            default='sub_rad',
            required=False)
                
ewoc_parser.add_argument("--plot_pid_pt", 
            action=plot_pid_pT,
            help="Optional argument which histograms the pT of particles "
                 +"with given particle ID that appear in the given "
                 +"process."\
                 +"\nRelies on input from the"\
                 +"```--write_pid_pt``` option of "
                 +"```write_ewocs```;",
            default=None,
            required=False)

ewoc_parser.add_argument("--plot_jet_pt", nargs=0, 
            action=plot_jet_pT,
            help="Optional argument which histograms the pT of jets "
                 +"with given parameters in the given process."\
                 +"\nRelies on input from the"\
                 +"```--write_jet_pt``` option of "
                 +"```write_ewocs```;",
            required=False)

ewoc_parser.add_argument("--plot_soft_energy", 
            action=plot_soft_energy, nargs=0,
            help="Optional argument which histograms the energy "\
                  "carried by the softer subjet in every pair of"\
                  "subjets found using the specified jet parameters;",
            default=None,
            required=False)


def str2bool(v):
    if isinstance(v, bool):
        return v
    if v.lower() in ('yes', 'true', 't', 'y', '1'):
        return True
    elif v.lower() in ('no', 'false', 'f', 'n', '0'):
        return False
    else:
        raise argparse.ArgumentTypeError('Boolean value expected.')

ewoc_parser.add_argument("--overwrite_hists",
            type=str2bool, nargs='?',
            const=True, default=False,
            dest="overwrite", 
            help="Flag determining whether to overwrite existing data"
                +" when generating plots;")
ewoc_parser.set_defaults(feature=True)

ewoc_parser.add_argument("--show_plots", 
            type=str2bool, nargs='?',
            const=True, default=True,
            dest="show", 
            help="Flag determining whether to show plots;")
